# Data analysis with Python

## Homework 1

In [103]:
import numpy as np
import pandas as pd
import os

In [104]:
os.getcwd()
os.chdir(r'E:\EMK data analysis')

In [105]:
os.getcwd()

'E:\\EMK data analysis'

#### 1. import the test_cricket.xlsx file in your jupyter notebook. Read the sheet 'wickets' from there. The dataset will be found here: https://github.com/SKawsar/Data_Analysis_with_Python


In [145]:
df = pd.read_excel("test_cricket.xlsx", sheet_name='wickets')
display(df)

Player       Span  Mat  Inns  Balls   Runs  Wkts  \
0   M Muralitharan (ICC/SL)  1992-2010  133   230  44039  18180   800   
1            SK Warne (AUS)  1992-2007  145   273  40705  17995   708   
2          A Kumble (INDIA)  1990-2008  132   236  40850  18355   619   
3         JM Anderson (ENG)  2003-2021  162   301  34791  16457   617   
4          GD McGrath (AUS)  1993-2007  124   243  29248  12186   563   
..                      ...        ...  ...   ...    ...    ...   ...   
74        SCG MacGill (AUS)  1998-2008   44    85  11237   6038   208   
75    Saqlain Mushtaq (PAK)  1995-2004   49    86  14070   6206   208   
76         AME Roberts (WI)  1974-1983   47    90  11135   5174   202   
77            JA Snow (ENG)  1965-1976   49    93  12021   5387   202   
78         JR Thomson (AUS)  1972-1985   51    90  10535   5601   200   

                    BBI                  BBM    Ave  Econ    SR   5  10  
0   1951-09-01 00:00:00               16/220  22.72  2.47  55.0  67  22  
1   1971-08-01 00:00:00               12/128  25.41  2.65  57.4  37  10  
2   1974-10-01 00:00:00               14/149  29.65  2.69  65.9  35   8  
3   1942-07-01 00:00:00  1971-11-01 00:00:00  26.67  2.83  56.3  30   3  
4   2021-08-24 00:00:00  2021-10-27 00:00:00  21.64  2.49  51.9  29   3  
..                  ...                  ...    ...   ...   ...  ..  ..  
74                8/108               12/107  29.02  3.22  54.0  12   2  
75                8/164               10/155  29.83  2.64  67.6  13   3  
76  1954-07-01 00:00:00               12/121  25.61  2.78  55.1  11   2  
77  1940-07-01 00:00:00               10/142  26.66  2.68  59.5   8   1  
78  1946-06-01 00:00:00                9/105  28.00  3.18  52.6   8   0  

[79 rows x 14 columns]

#### 2. Display the first 10 rows of the dataframe

In [146]:
display(df.head(10))

Player       Span  Mat  Inns  Balls   Runs  Wkts  \
0  M Muralitharan (ICC/SL)  1992-2010  133   230  44039  18180   800   
1           SK Warne (AUS)  1992-2007  145   273  40705  17995   708   
2         A Kumble (INDIA)  1990-2008  132   236  40850  18355   619   
3        JM Anderson (ENG)  2003-2021  162   301  34791  16457   617   
4         GD McGrath (AUS)  1993-2007  124   243  29248  12186   563   
5          SCJ Broad (ENG)  2007-2021  148   272  29713  14502   523   
6            CA Walsh (WI)  1984-2001  132   242  30019  12688   519   
7            DW Steyn (SA)  2004-2019   93   171  18608  10077   439   
8      N Kapil Dev (INDIA)  1978-1994  131   227  27740  12867   434   
9        HMRKB Herath (SL)  1999-2018   93   170  25993  12157   433   

                   BBI                  BBM    Ave  Econ    SR   5  10  
0  1951-09-01 00:00:00               16/220  22.72  2.47  55.0  67  22  
1  1971-08-01 00:00:00               12/128  25.41  2.65  57.4  37  10  
2  1974-10-01 00:00:00               14/149  29.65  2.69  65.9  35   8  
3  1942-07-01 00:00:00  1971-11-01 00:00:00  26.67  2.83  56.3  30   3  
4  2021-08-24 00:00:00  2021-10-27 00:00:00  21.64  2.49  51.9  29   3  
5  2021-08-15 00:00:00               11/121  27.72  2.92  56.8  18   3  
6  1937-07-01 00:00:00                13/55  24.44  2.53  57.8  22   3  
7  1951-07-01 00:00:00  1960-11-01 00:00:00  22.95  3.24  42.3  26   5  
8  1983-09-01 00:00:00               11/146  29.64  2.78  63.9  23   2  
9                9/127               14/184  28.07  2.80  60.0  34   9

#### 3. Create a markdown cell and explain the meaning of each column

##### {'Player'= 'Player_Name', 'Span'= 'Career_duration', 'Mat'= 'No_match_played', 'Inns' = 'No_innings', 'Balls':'No_balls_bowled','Runs': 'Runs_considered', 'Wkts' = 'Total_wickets_taken', 'BBI' = 'Best_Bowling_in_Innings', 'BBM' = 'Best_Bowling_in_Match','Ave' = 'Average_runs_per_wicket', 'Econ' = 'average_runs_conceded_per_over', 'SR'= 'Strike_rate', 5 = 'No_times_5_wickets_taken', 10 = 'No_times_10_wickets_taken'}

#### 4. find the number of rows and columns in the dataframe

In [147]:
print("number of rows = ", df.shape[0])

number of rows =  79


In [148]:
print("number of columns = ", df.shape[1])

number of columns =  14


#### 5. find the data statistics and check for the data types

In [149]:
display(df.describe())

Mat        Inns         Balls          Runs        Wkts  \
count   79.000000   79.000000     79.000000     79.000000   79.000000   
mean    80.101266  144.797468  18630.303797   8595.506329  317.101266   
std     28.537692   51.042310   7190.036515   3080.256645  121.731587   
min     37.000000   67.000000   8785.000000   4846.000000  200.000000   
25%     60.500000  110.000000  13580.000000   6456.500000  229.000000   
50%     71.000000  129.000000  16498.000000   7742.000000  266.000000   
75%     93.000000  169.000000  21742.500000   9756.000000  374.500000   
max    166.000000  301.000000  44039.000000  18355.000000  800.000000   

             Ave       Econ         SR          5         10  
count  79.000000  79.000000  79.000000  79.000000  79.000000  
mean   27.466456   2.806582  59.187342  16.354430   2.797468  
std     3.657561   0.351666   9.349337   9.642372   3.235935  
min    20.940000   1.980000  41.200000   3.000000   0.000000  
25%    24.425000   2.600000  53.300000   9.500000   1.000000  
50%    28.000000   2.820000  57.400000  14.000000   2.000000  
75%    29.870000   3.080000  63.950000  20.500000   3.500000  
max    34.790000   3.460000  91.900000  67.000000  22.000000

In [150]:
print(df.dtypes)

Player     object
Span       object
Mat         int64
Inns        int64
Balls       int64
Runs        int64
Wkts        int64
BBI        object
BBM        object
Ave       float64
Econ      float64
SR        float64
5           int64
10          int64
dtype: object


#### 6. Are there any missing values present in the dataset?

###### Ans: No,there are no missing value in the dataset. Because there are no column which shows null. All shows non-null. And all is false. False means not null

In [151]:
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 79 entries, 0 to 78
Data columns (total 14 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Player  79 non-null     object 
 1   Span    79 non-null     object 
 2   Mat     79 non-null     int64  
 3   Inns    79 non-null     int64  
 4   Balls   79 non-null     int64  
 5   Runs    79 non-null     int64  
 6   Wkts    79 non-null     int64  
 7   BBI     79 non-null     object 
 8   BBM     79 non-null     object 
 9   Ave     79 non-null     float64
 10  Econ    79 non-null     float64
 11  SR      79 non-null     float64
 12  5       79 non-null     int64  
 13  10      79 non-null     int64  
dtypes: float64(3), int64(7), object(4)
memory usage: 8.8+ KB
None


In [152]:
print(df.isnull())

    Player   Span    Mat   Inns  Balls   Runs   Wkts    BBI    BBM    Ave  \
0    False  False  False  False  False  False  False  False  False  False   
1    False  False  False  False  False  False  False  False  False  False   
2    False  False  False  False  False  False  False  False  False  False   
3    False  False  False  False  False  False  False  False  False  False   
4    False  False  False  False  False  False  False  False  False  False   
..     ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   
74   False  False  False  False  False  False  False  False  False  False   
75   False  False  False  False  False  False  False  False  False  False   
76   False  False  False  False  False  False  False  False  False  False   
77   False  False  False  False  False  False  False  False  False  False   
78   False  False  False  False  False  False  False  False  False  False   

     Econ     SR      5     10  
0   False  False  False  False  
1   False

#### 7. Rename the column names appropriately

In [153]:
df = df.rename(columns = {'Player':'Player_Name',
                          'Span':'Career_duration',
                          'Mat':'No_match_played', 
                          'Inns':'No_innings', 
                          'Balls':'No_balls_bowled',
                          'Runs': 'Runs_considered', 
                          'Wkts':'Total_wickets_taken', 
                          'BBI': 'Best_Bowling_in_Innings', 
                          'BBM':'Best_Bowling_in_Match',
                          'Ave':'Average_runs_per_wicket',
                          'Econ':'average_runs_conceded_per_over', 
                          'SR':'Strike_rate',
                          5: 'No_times_5_wickets_taken', 
                          10:'No_times_10_wickets_taken'})
display(df.head())


Player_Name Career_duration  No_match_played  No_innings  \
0  M Muralitharan (ICC/SL)       1992-2010              133         230   
1           SK Warne (AUS)       1992-2007              145         273   
2         A Kumble (INDIA)       1990-2008              132         236   
3        JM Anderson (ENG)       2003-2021              162         301   
4         GD McGrath (AUS)       1993-2007              124         243   

   No_balls_bowled  Runs_considered  Total_wickets_taken  \
0            44039            18180                  800   
1            40705            17995                  708   
2            40850            18355                  619   
3            34791            16457                  617   
4            29248            12186                  563   

  Best_Bowling_in_Innings Best_Bowling_in_Match  Average_runs_per_wicket  \
0     1951-09-01 00:00:00                16/220                    22.72   
1     1971-08-01 00:00:00                12/128                    25.41   
2     1974-10-01 00:00:00                14/149                    29.65   
3     1942-07-01 00:00:00   1971-11-01 00:00:00                    26.67   
4     2021-08-24 00:00:00   2021-10-27 00:00:00                    21.64   

   average_runs_conceded_per_over  Strike_rate  No_times_5_wickets_taken  \
0                            2.47         55.0                        67   
1                            2.65         57.4                        37   
2                            2.69         65.9                        35   
3                            2.83         56.3                        30   
4                            2.49         51.9                        29   

   No_times_10_wickets_taken  
0                         22  
1                         10  
2                          8  
3                          3  
4                          3

#### 8.Remove a column from the dataframe

In [154]:
df = df.drop('Player_Name', axis=1)
display(df.head())


Career_duration  No_match_played  No_innings  No_balls_bowled  \
0       1992-2010              133         230            44039   
1       1992-2007              145         273            40705   
2       1990-2008              132         236            40850   
3       2003-2021              162         301            34791   
4       1993-2007              124         243            29248   

   Runs_considered  Total_wickets_taken Best_Bowling_in_Innings  \
0            18180                  800     1951-09-01 00:00:00   
1            17995                  708     1971-08-01 00:00:00   
2            18355                  619     1974-10-01 00:00:00   
3            16457                  617     1942-07-01 00:00:00   
4            12186                  563     2021-08-24 00:00:00   

  Best_Bowling_in_Match  Average_runs_per_wicket  \
0                16/220                    22.72   
1                12/128                    25.41   
2                14/149                    29.65   
3   1971-11-01 00:00:00                    26.67   
4   2021-10-27 00:00:00                    21.64   

   average_runs_conceded_per_over  Strike_rate  No_times_5_wickets_taken  \
0                            2.47         55.0                        67   
1                            2.65         57.4                        37   
2                            2.69         65.9                        35   
3                            2.83         56.3                        30   
4                            2.49         51.9                        29   

   No_times_10_wickets_taken  
0                         22  
1                         10  
2                          8  
3                          3  
4                          3

### This is the end of HW1

## Data analysis with Python
## Homework 2

In [192]:
import numpy as np
import pandas as pd
import os

In [193]:
os.getcwd()
os.chdir(r'E:\EMK data analysis')

In [229]:
df_wickets = pd.read_csv("wickets.csv")
display(df_wickets)

Player \tSpan \tMat \tInns \tBalls \tRuns \tWkts \tBBI \tBBM \tAve \tEcon \tSR \t5 \t10
0   M Muralitharan (ICC/SL) \t1992-2010 \t133 \t23...                                     
1   SK Warne (AUS) \t1992-2007 \t145 \t273 \t40705...                                     
2   JM Anderson (ENG) \t2003-2021 \t164* \t304 \t3...                                     
3   A Kumble (INDIA) \t1990-2008 \t132 \t236 \t408...                                     
4   GD McGrath (AUS) \t1993-2007 \t124 \t243 \t292...                                     
..                                                ...                                     
74  SCG MacGill (AUS) \t1998-2008 \t44 \t85 \t1123...                                     
75  Saqlain Mushtaq (PAK) \t1995-2004 \t49 \t86 \t...                                     
76  AME Roberts (WI) \t1974-1983 \t47 \t90 \t11135...                                     
77  JA Snow (ENG) \t1965-1976 \t49 \t93 \t12021 \t...                                     
78  JR Thomson (AUS) \t1972-1985 \t51 \t90 \t10535...                                     

[79 rows x 1 columns]

In [230]:
df_wickets = pd.read_csv('wickets.csv', encoding='unicode_escape', delimiter='\t')
display(df_wickets)


Player        Span    Mat   Inns   Balls   Runs   Wkts   \
0   M Muralitharan (ICC/SL)   1992-2010    133     230   44039  18180    800   
1            SK Warne (AUS)   1992-2007    145     273   40705  17995    708   
2         JM Anderson (ENG)   2003-2021   164*     304   35079  16575    623   
3          A Kumble (INDIA)   1990-2008    132     236   40850  18355    619   
4          GD McGrath (AUS)   1993-2007    124     243   29248  12186    563   
..                       ...         ...    ...    ...     ...    ...    ...   
74        SCG MacGill (AUS)   1998-2008     44      85   11237   6038    208   
75    Saqlain Mushtaq (PAK)   1995-2004     49      86   14070   6206    208   
76         AME Roberts (WI)   1974-1983     47      90   11135   5174    202   
77            JA Snow (ENG)   1965-1976     49      93   12021   5387    202   
78         JR Thomson (AUS)   1972-1985     51      90   10535   5601    200   

      BBI      BBM    Ave   Econ    SR   5   10  
0    9/51   16/220   22.72   2.47  55.0  67  22  
1    8/71   12/128   25.41   2.65  57.4  37  10  
2    7/42    11/71   26.60   2.83  56.3  30   3  
3   10/74   14/149   29.65   2.69  65.9  35   8  
4    8/24    10/27   21.64   2.49  51.9  29   3  
..     ...      ...    ...    ...   ...  ..  ..  
74  8/108   12/107   29.02   3.22  54.0  12   2  
75  8/164   10/155   29.83   2.64  67.6  13   3  
76   7/54   12/121   25.61   2.78  55.1  11   2  
77   7/40   10/142   26.66   2.68  59.5   8   1  
78   6/46    9/105   28.00   3.18  52.6   8   0  

[79 rows x 14 columns]

In [231]:
display(df_wickets.head())

Player        Span    Mat   Inns   Balls   Runs   Wkts   \
0  M Muralitharan (ICC/SL)   1992-2010    133     230   44039  18180    800   
1           SK Warne (AUS)   1992-2007    145     273   40705  17995    708   
2        JM Anderson (ENG)   2003-2021   164*     304   35079  16575    623   
3         A Kumble (INDIA)   1990-2008    132     236   40850  18355    619   
4         GD McGrath (AUS)   1993-2007    124     243   29248  12186    563   

     BBI      BBM    Ave   Econ    SR   5   10  
0   9/51   16/220   22.72   2.47  55.0  67  22  
1   8/71   12/128   25.41   2.65  57.4  37  10  
2   7/42    11/71   26.60   2.83  56.3  30   3  
3  10/74   14/149   29.65   2.69  65.9  35   8  
4   8/24    10/27   21.64   2.49  51.9  29   3

In [232]:
print(df_wickets.columns)

Index(['Player ', 'Span ', 'Mat ', 'Inns ', 'Balls ', 'Runs ', 'Wkts ', 'BBI ',
       'BBM ', 'Ave ', 'Econ ', 'SR ', '5 ', '10'],
      dtype='object')


### 1. Remove the column BBI, BBM

In [233]:
df_wickets = df_wickets.drop(['BBI ','BBM '], axis=1)
display(df_wickets.head())

Player        Span    Mat   Inns   Balls   Runs   Wkts   \
0  M Muralitharan (ICC/SL)   1992-2010    133     230   44039  18180    800   
1           SK Warne (AUS)   1992-2007    145     273   40705  17995    708   
2        JM Anderson (ENG)   2003-2021   164*     304   35079  16575    623   
3         A Kumble (INDIA)   1990-2008    132     236   40850  18355    619   
4         GD McGrath (AUS)   1993-2007    124     243   29248  12186    563   

    Ave   Econ    SR   5   10  
0  22.72   2.47  55.0  67  22  
1  25.41   2.65  57.4  37  10  
2  26.60   2.83  56.3  30   3  
3  29.65   2.69  65.9  35   8  
4  21.64   2.49  51.9  29   3

###### Removed two columns  'BBI' and 'BBM' from dataset

In [234]:
print(df_wickets.columns)

Index(['Player ', 'Span ', 'Mat ', 'Inns ', 'Balls ', 'Runs ', 'Wkts ', 'Ave ',
       'Econ ', 'SR ', '5 ', '10'],
      dtype='object')


##### Removed two columns 'BBI' and 'BBM' from dataset

### 2. How many players played for ICC?

In [235]:
df_player = df_wickets['Player '].str.split("(", expand=True)

display(df_player.head(5))

0         1
0  M Muralitharan   ICC/SL) 
1        SK Warne      AUS) 
2     JM Anderson      ENG) 
3        A Kumble    INDIA) 
4      GD McGrath      AUS)

#### Splitting the 'Player' column to get the information about 'Country'

In [236]:
df_wickets = pd.concat([df_wickets, df_player], axis=1)

display(df_wickets.head())

Player        Span    Mat   Inns   Balls   Runs   Wkts   \
0  M Muralitharan (ICC/SL)   1992-2010    133     230   44039  18180    800   
1           SK Warne (AUS)   1992-2007    145     273   40705  17995    708   
2        JM Anderson (ENG)   2003-2021   164*     304   35079  16575    623   
3         A Kumble (INDIA)   1990-2008    132     236   40850  18355    619   
4         GD McGrath (AUS)   1993-2007    124     243   29248  12186    563   

    Ave   Econ    SR   5   10                0         1  
0  22.72   2.47  55.0  67  22  M Muralitharan   ICC/SL)   
1  25.41   2.65  57.4  37  10        SK Warne      AUS)   
2  26.60   2.83  56.3  30   3     JM Anderson      ENG)   
3  29.65   2.69  65.9  35   8        A Kumble    INDIA)   
4  21.64   2.49  51.9  29   3      GD McGrath      AUS)

##### Concatinating the new dataframe with the main dataframe

In [237]:
df_wickets = df_wickets.drop('Player ', axis=1)

display(df_wickets.head())

Span    Mat   Inns   Balls   Runs   Wkts    Ave   Econ    SR   5   10  \
0  1992-2010    133     230   44039  18180    800  22.72   2.47  55.0  67  22   
1  1992-2007    145     273   40705  17995    708  25.41   2.65  57.4  37  10   
2  2003-2021   164*     304   35079  16575    623  26.60   2.83  56.3  30   3   
3  1990-2008    132     236   40850  18355    619  29.65   2.69  65.9  35   8   
4  1993-2007    124     243   29248  12186    563  21.64   2.49  51.9  29   3   

                 0         1  
0  M Muralitharan   ICC/SL)   
1        SK Warne      AUS)   
2     JM Anderson      ENG)   
3        A Kumble    INDIA)   
4      GD McGrath      AUS)

###### Dropped the 'Player' column

In [238]:
df_wickets = df_wickets.rename(columns={0: 'Player',
                                        1: 'Country'})

display(df_wickets.head())

Span    Mat   Inns   Balls   Runs   Wkts    Ave   Econ    SR   5   10  \
0  1992-2010    133     230   44039  18180    800  22.72   2.47  55.0  67  22   
1  1992-2007    145     273   40705  17995    708  25.41   2.65  57.4  37  10   
2  2003-2021   164*     304   35079  16575    623  26.60   2.83  56.3  30   3   
3  1990-2008    132     236   40850  18355    619  29.65   2.69  65.9  35   8   
4  1993-2007    124     243   29248  12186    563  21.64   2.49  51.9  29   3   

            Player   Country  
0  M Muralitharan   ICC/SL)   
1        SK Warne      AUS)   
2     JM Anderson      ENG)   
3        A Kumble    INDIA)   
4      GD McGrath      AUS)

##### Renamed the column names as player and country

In [239]:
# remove the ")" from the 'Country' column
df_wickets['Country'] = df_wickets['Country'].str.replace(")","")

display(df_wickets.head())

<ipython-input-239-b9ca476b04f7>:2: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will*not* be treated as literal strings when regex=True.
  df_wickets['Country'] = df_wickets['Country'].str.replace(")","")


Span    Mat   Inns   Balls   Runs   Wkts    Ave   Econ    SR   5   10  \
0  1992-2010    133     230   44039  18180    800  22.72   2.47  55.0  67  22   
1  1992-2007    145     273   40705  17995    708  25.41   2.65  57.4  37  10   
2  2003-2021   164*     304   35079  16575    623  26.60   2.83  56.3  30   3   
3  1990-2008    132     236   40850  18355    619  29.65   2.69  65.9  35   8   
4  1993-2007    124     243   29248  12186    563  21.64   2.49  51.9  29   3   

            Player  Country  
0  M Muralitharan   ICC/SL   
1        SK Warne      AUS   
2     JM Anderson      ENG   
3        A Kumble    INDIA   
4      GD McGrath      AUS

In [240]:
new_col_sequence = ['Player', 'Country','Span ', 'Mat ', 'Inns ', 'Balls ', 'Runs ', 'Wkts ', 'Ave ', 'Econ ',
       'SR ', '5 ', '10']

In [241]:
df_wickets = df_wickets[new_col_sequence]

display(df_wickets.head())

Player  Country       Span    Mat   Inns   Balls   Runs   Wkts   \
0  M Muralitharan   ICC/SL   1992-2010    133     230   44039  18180    800   
1        SK Warne      AUS   1992-2007    145     273   40705  17995    708   
2     JM Anderson      ENG   2003-2021   164*     304   35079  16575    623   
3        A Kumble    INDIA   1990-2008    132     236   40850  18355    619   
4      GD McGrath      AUS   1993-2007    124     243   29248  12186    563   

    Ave   Econ    SR   5   10  
0  22.72   2.47  55.0  67  22  
1  25.41   2.65  57.4  37  10  
2  26.60   2.83  56.3  30   3  
3  29.65   2.69  65.9  35   8  
4  21.64   2.49  51.9  29   3

##### The columns are rearranged 

In [242]:
def icc_check(x):
    if "ICC" in x:
        return "Yes"
    else:
        return "No"

In [243]:
df_wickets['played_for_ICC'] = df_wickets['Country'].apply(icc_check)

display(df_wickets.head(10))

Player  Country       Span    Mat   Inns   Balls   Runs   Wkts   \
0  M Muralitharan   ICC/SL   1992-2010    133     230   44039  18180    800   
1        SK Warne      AUS   1992-2007    145     273   40705  17995    708   
2     JM Anderson      ENG   2003-2021   164*     304   35079  16575    623   
3        A Kumble    INDIA   1990-2008    132     236   40850  18355    619   
4      GD McGrath      AUS   1993-2007    124     243   29248  12186    563   
5       SCJ Broad      ENG   2007-2021    149     274   29863  14590    524   
6        CA Walsh       WI   1984-2001    132     242   30019  12688    519   
7        DW Steyn       SA   2004-2019     93     171   18608  10077    439   
8     N Kapil Dev    INDIA   1978-1994    131     227   27740  12867    434   
9    HMRKB Herath       SL   1999-2018     93     170   25993  12157    433   

    Ave   Econ    SR   5   10 played_for_ICC  
0  22.72   2.47  55.0  67  22            Yes  
1  25.41   2.65  57.4  37  10             No  
2  26.60   2.83  56.3  30   3             No  
3  29.65   2.69  65.9  35   8             No  
4  21.64   2.49  51.9  29   3             No  
5  27.84   2.93  56.9  18   3             No  
6  24.44   2.53  57.8  22   3             No  
7  22.95   3.24  42.3  26   5             No  
8  29.64   2.78  63.9  23   2             No  
9  28.07   2.80  60.0  34   9             No

##### Checked that a player had played for ICC or not

In [244]:
df_wickets['played_for_ICC'].value_counts()

No     74
Yes     5
Name: played_for_ICC, dtype: int64

##### Showing how many players played for ICC

In [245]:
df_wickets['Country'] = df_wickets['Country'].str.replace("ICC/", "")

display(df_wickets.head())

Player Country       Span    Mat   Inns   Balls   Runs   Wkts   \
0  M Muralitharan      SL   1992-2010    133     230   44039  18180    800   
1        SK Warne     AUS   1992-2007    145     273   40705  17995    708   
2     JM Anderson     ENG   2003-2021   164*     304   35079  16575    623   
3        A Kumble   INDIA   1990-2008    132     236   40850  18355    619   
4      GD McGrath     AUS   1993-2007    124     243   29248  12186    563   

    Ave   Econ    SR   5   10 played_for_ICC  
0  22.72   2.47  55.0  67  22            Yes  
1  25.41   2.65  57.4  37  10             No  
2  26.60   2.83  56.3  30   3             No  
3  29.65   2.69  65.9  35   8             No  
4  21.64   2.49  51.9  29   3             No

##### To get the country name, 'ICC/' is removed from the dataset 

### 3. How many different countries are present in this dataset? 

In [246]:
df_wickets['Country'].value_counts()

AUS         18
ENG         13
INDIA       10
WI           9
SA           8
PAK          7
NZ           7
SL           3
ENG/ICC      2
BDESH        1
ZIM          1
Name: Country, dtype: int64

##### There are 11 different type countries in the dataset

### 4. Which player(s) had played for the longest period of time?

In [248]:
df_span = df_wickets['Span '].str.split("-", expand=True)
display(df_wickets.head(5))

Player Country       Span    Mat   Inns   Balls   Runs   Wkts   \
0  M Muralitharan      SL   1992-2010    133     230   44039  18180    800   
1        SK Warne     AUS   1992-2007    145     273   40705  17995    708   
2     JM Anderson     ENG   2003-2021   164*     304   35079  16575    623   
3        A Kumble   INDIA   1990-2008    132     236   40850  18355    619   
4      GD McGrath     AUS   1993-2007    124     243   29248  12186    563   

    Ave   Econ    SR   5   10 played_for_ICC  
0  22.72   2.47  55.0  67  22            Yes  
1  25.41   2.65  57.4  37  10             No  
2  26.60   2.83  56.3  30   3             No  
3  29.65   2.69  65.9  35   8             No  
4  21.64   2.49  51.9  29   3             No

##### splitted the 'Span' column based on the "-"

In [249]:
df_wickets = pd.concat([df_wickets, df_span], axis=1)
display(df_wickets.head(5))

Player Country       Span    Mat   Inns   Balls   Runs   Wkts   \
0  M Muralitharan      SL   1992-2010    133     230   44039  18180    800   
1        SK Warne     AUS   1992-2007    145     273   40705  17995    708   
2     JM Anderson     ENG   2003-2021   164*     304   35079  16575    623   
3        A Kumble   INDIA   1990-2008    132     236   40850  18355    619   
4      GD McGrath     AUS   1993-2007    124     243   29248  12186    563   

    Ave   Econ    SR   5   10 played_for_ICC     0      1  
0  22.72   2.47  55.0  67  22            Yes  1992  2010   
1  25.41   2.65  57.4  37  10             No  1992  2007   
2  26.60   2.83  56.3  30   3             No  2003  2021   
3  29.65   2.69  65.9  35   8             No  1990  2008   
4  21.64   2.49  51.9  29   3             No  1993  2007

##### Concatinated the new dataframe with the main dataframe

In [250]:
df_wickets = df_wickets.rename(columns={0: "start_year",
                                        1: "end_year"})
display(df_wickets.head(5))

Player Country       Span    Mat   Inns   Balls   Runs   Wkts   \
0  M Muralitharan      SL   1992-2010    133     230   44039  18180    800   
1        SK Warne     AUS   1992-2007    145     273   40705  17995    708   
2     JM Anderson     ENG   2003-2021   164*     304   35079  16575    623   
3        A Kumble   INDIA   1990-2008    132     236   40850  18355    619   
4      GD McGrath     AUS   1993-2007    124     243   29248  12186    563   

    Ave   Econ    SR   5   10 played_for_ICC start_year end_year  
0  22.72   2.47  55.0  67  22            Yes       1992    2010   
1  25.41   2.65  57.4  37  10             No       1992    2007   
2  26.60   2.83  56.3  30   3             No       2003    2021   
3  29.65   2.69  65.9  35   8             No       1990    2008   
4  21.64   2.49  51.9  29   3             No       1993    2007

##### Renamed the newly created column names to start_year and end_year

In [251]:
df_wickets = df_wickets.drop("Span ", axis=1)

display(df_wickets.head())

Player Country   Mat   Inns   Balls   Runs   Wkts    Ave   Econ   \
0  M Muralitharan      SL    133     230   44039  18180    800  22.72   2.47   
1        SK Warne     AUS    145     273   40705  17995    708  25.41   2.65   
2     JM Anderson     ENG   164*     304   35079  16575    623  26.60   2.83   
3        A Kumble   INDIA    132     236   40850  18355    619  29.65   2.69   
4      GD McGrath     AUS    124     243   29248  12186    563  21.64   2.49   

    SR   5   10 played_for_ICC start_year end_year  
0  55.0  67  22            Yes       1992    2010   
1  57.4  37  10             No       1992    2007   
2  56.3  30   3             No       2003    2021   
3  65.9  35   8             No       1990    2008   
4  51.9  29   3             No       1993    2007

##### Dropped the "Span" column because two new columns are created mentioning the information

In [252]:
df_wickets.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 79 entries, 0 to 78
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Player          79 non-null     object 
 1   Country         79 non-null     object 
 2   Mat             79 non-null     object 
 3   Inns            79 non-null     int64  
 4   Balls           79 non-null     int64  
 5   Runs            79 non-null     int64  
 6   Wkts            79 non-null     int64  
 7   Ave             79 non-null     float64
 8   Econ            79 non-null     float64
 9   SR              79 non-null     float64
 10  5               79 non-null     int64  
 11  10              79 non-null     int64  
 12  played_for_ICC  79 non-null     object 
 13  start_year      79 non-null     object 
 14  end_year        79 non-null     object 
dtypes: float64(3), int64(6), object(6)
memory usage: 9.4+ KB


##### Data type of start_year and end_year are object. We need those as integer

In [253]:
df_wickets['start_year'] = df_wickets['start_year'].astype('int') 
df_wickets['end_year'] = df_wickets['end_year'].astype('int')

print(df_wickets.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 79 entries, 0 to 78
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Player          79 non-null     object 
 1   Country         79 non-null     object 
 2   Mat             79 non-null     object 
 3   Inns            79 non-null     int64  
 4   Balls           79 non-null     int64  
 5   Runs            79 non-null     int64  
 6   Wkts            79 non-null     int64  
 7   Ave             79 non-null     float64
 8   Econ            79 non-null     float64
 9   SR              79 non-null     float64
 10  5               79 non-null     int64  
 11  10              79 non-null     int64  
 12  played_for_ICC  79 non-null     object 
 13  start_year      79 non-null     int32  
 14  end_year        79 non-null     int32  
dtypes: float64(3), int32(2), int64(6), object(4)
memory usage: 8.8+ KB
None


##### Data type of start_year and end_year are integer which is required to get years_played

In [254]:
df_wickets['years_played'] = df_wickets['end_year'] - df_wickets['start_year']

df_wickets= df_wickets.drop(['start_year', "end_year"], axis=1)

display(df_wickets.head(10))

Player Country   Mat   Inns   Balls   Runs   Wkts    Ave   Econ   \
0  M Muralitharan      SL    133     230   44039  18180    800  22.72   2.47   
1        SK Warne     AUS    145     273   40705  17995    708  25.41   2.65   
2     JM Anderson     ENG   164*     304   35079  16575    623  26.60   2.83   
3        A Kumble   INDIA    132     236   40850  18355    619  29.65   2.69   
4      GD McGrath     AUS    124     243   29248  12186    563  21.64   2.49   
5       SCJ Broad     ENG    149     274   29863  14590    524  27.84   2.93   
6        CA Walsh      WI    132     242   30019  12688    519  24.44   2.53   
7        DW Steyn      SA     93     171   18608  10077    439  22.95   3.24   
8     N Kapil Dev   INDIA    131     227   27740  12867    434  29.64   2.78   
9    HMRKB Herath      SL     93     170   25993  12157    433  28.07   2.80   

    SR   5   10 played_for_ICC  years_played  
0  55.0  67  22            Yes            18  
1  57.4  37  10             No            15  
2  56.3  30   3             No            18  
3  65.9  35   8             No            18  
4  51.9  29   3             No            14  
5  56.9  18   3             No            14  
6  57.8  22   3             No            17  
7  42.3  26   5             No            15  
8  63.9  23   2             No            16  
9  60.0  34   9             No            19

##### Got the data mentioning years_played

In [255]:
df_wickets.sort_values(by='years_played', ascending = False).head(5)

Player Country   Mat   Inns   Balls   Runs   Wkts    Ave   Econ   \
21      Imran Khan     PAK     88     142   19458   8258    362  22.81   2.54   
55       GS Sobers      WI     93     159   21599   7999    235  34.03   2.22   
9     HMRKB Herath      SL     93     170   25993  12157    433  28.07   2.80   
0   M Muralitharan      SL    133     230   44039  18180    800  22.72   2.47   
2      JM Anderson     ENG   164*     304   35079  16575    623  26.60   2.83   

     SR   5   10 played_for_ICC  years_played  
21  53.7  23   6             No            21  
55  91.9   6   0             No            20  
9   60.0  34   9             No            19  
0   55.0  67  22            Yes            18  
2   56.3  30   3             No            18

##### Imran Khan had played for the longest period of time i.e 21 years

### 5. Which player(s) had played for the shortest period of time?

In [256]:
df_wickets.sort_values(by='years_played', ascending = False).tail(5)

Player   Country  Mat   Inns   Balls   Runs   Wkts    Ave   Econ   \
54    Yasir Shah       PAK   46*      84   13607   7248    235  30.84   3.19   
61   SJ Harmison   ENG/ICC    63     115   13375   7192    226  31.82   3.22   
72  JR Hazlewood       AUS    55     103   11887   5438    212  25.65   2.74   
71      K Rabada        SA    47      86    8785   4846    213  22.75   3.30   
44      GP Swann       ENG    60     109   15349   7642    255  29.96   2.98   

     SR   5   10 played_for_ICC  years_played  
54  57.9  16   3             No             7  
61  59.1   8   1            Yes             7  
72  56.0   9   0             No             7  
71  41.2  10   4             No             6  
44  60.1  17   3             No             5

##### GP Swann had played for the shortest period of time i.e 5 years

### 6. How many Australian Bowlers are present in this dataset?

In [258]:
df_wickets['Country'].value_counts()

AUS         18
ENG         13
INDIA       10
WI           9
SA           8
PAK          7
NZ           7
SL           3
ENG/ICC      2
BDESH        1
ZIM          1
Name: Country, dtype: int64

##### The dataset displayed the number of bowlers from different countries
##### There are 18 Australian bowlers in the dataset

### 7. Is there any Bangladeshi player present in this dataset?
##### Yes, there are only one Bangladeshi player present in this dataset

### 8. Which player had the lowest economy rate?

In [259]:
df_wickets.sort_values(by='Econ ', ascending = False).tail(5)

Player Country Mat   Inns   Balls   Runs   Wkts    Ave   Econ   \
68   CV Grimmett     AUS   37      67   14513   5231    216  24.21   2.16   
39       BS Bedi   INDIA   67     118   21364   7637    266  28.71   2.14   
47      R Benaud     AUS   63     116   19108   6704    248  27.03   2.10   
35  DL Underwood     ENG   86     151   21862   7674    297  25.83   2.10   
32      LR Gibbs      WI   79     148   27115   8989    309  29.09   1.98   

     SR   5   10 played_for_ICC  years_played  
68  67.1  21   7             No            11  
39  80.3  14   1             No            13  
47  77.0  16   1             No            12  
35  73.6  17   6             No            16  
32  87.7  18   2             No            18

##### LR Gibbs had the lowest economy rate (1.98)

### 9. Which player had the lowest strike rate?

In [260]:
df_wickets.sort_values(by='SR ', ascending = False).tail(5)

Player Country Mat   Inns   Balls   Runs   Wkts    Ave   Econ   \
25     AA Donald      SA   72     129   15519   7344    330  22.25   2.83   
19   MD Marshall      WI   81     151   17584   7876    376  20.94   2.68   
20  Waqar Younis     PAK   87     154   16224   8788    373  23.56   3.25   
7       DW Steyn      SA   93     171   18608  10077    439  22.95   3.24   
71      K Rabada      SA   47      86    8785   4846    213  22.75   3.30   

     SR   5   10 played_for_ICC  years_played  
25  47.0  20   3             No            10  
19  46.7  22   4             No            13  
20  43.4  22   5             No            14  
7   42.3  26   5             No            15  
71  41.2  10   4             No             6

##### K Rabada had the lowest strike rate (41.2)

### 10. Which player had the lowest bowling average?

In [223]:
df_wickets.sort_values(by='Ave ', ascending = False).tail(5)

Player Country  Mat   Inns   Balls   Runs   Wkts    Ave   Econ   \
4    GD McGrath     AUS   124     243   29248  12186    563  21.64   2.49   
33   FS Trueman     ENG    67     127   15178   6625    307  21.57   2.61   
15  CEL Ambrose      WI    98     179   22103   8501    405  20.99   2.30   
41     J Garner      WI    58     111   13169   5433    259  20.97   2.47   
19  MD Marshall      WI    81     151   17584   7876    376  20.94   2.68   

     SR   5   10 played_for_ICC  years_played  
4   51.9  29   3             No            14  
33  49.4  17   3             No            13  
15  54.5  22   3             No            12  
41  50.8   7   0             No            10  
19  46.7  22   4             No            13

##### MD Marshall had the lowest bowling average (20.94)

### This is the end of Homework 2